File: Algorithm_Final_Project/overlay_blocks.cpp
/*
Simple algorithm project for university
By NimaH79
NimaH79.ir
**/

#include <bits/stdc++.h>
using namespace std;
 
struct Rectangle 
{
    int x_start;
    int height;
    int x_end;
}; 

bool compareTwoRectangles(Rectangle a, Rectangle b)
{
    if(a.x_start != b.x_start)
        return a.x_start < b.x_start;

    if(a.x_end != b.x_end)
        return a.x_end < b.x_end;
 
    return a.height < b.height;
}

void sortRectangles(Rectangle a[], int n)
{
    sort(a, a + n, compareTwoRectangles);
}

int main()
{
    int n, result[1000], result_count = 0, previous_height = 0, max_height;

    cin >> n;

    Rectangle rectangles[n];

    for(int i = 0; i < n; i++) {
        cin >> rectangles[i].x_start >> rectangles[i].height >> rectangles[i].x_end;
    }
 
    sortRectangles(rectangles, n);

    for(int i = rectangles[0].x_start; i < rectangles[n - 1].x_end; i++)
    {
        max_height = 0;
        for(int j = 0; j < n; j++)
        {
            if(rectangles[j].x_start <= i && rectangles[j].x_end > i && rectangles[j].height > max_height)
                max_height = rectangles[j].height;
        }
        if(max_height != previous_height) {
            result[result_count] = i;
            result_count++;
            result[result_count] = max_height;
            result_count++;
        }
        previous_height = max_height;
    }

    result[result_count] = rectangles[n - 1].x_end;
    result_count++;

    for(int i = 0; i < result_count; i++) {
        if (i > 0)
            cout << ",";
        cout << result[i];
    }
 
    return 0;
}

File: Algorithm_Final_Project/overlay_blocks.py
# Simple algorithm project for university
# By NimaH79
# NimaH79.ir

from operator import itemgetter

rectangles = []
while 1:
    rectangle = input()
    if rectangle == 'finish':
        break
    rectangle = rectangle.split(',')
    if len(rectangle) % 2 == 0:
        while len(rectangle) >= 3:
            a = int(rectangle[0])
            b = int(rectangle[1])
            c = int(rectangle[2])
            while a < c:
                rectangles.append([a, b, a + 1])
                a += 1
            rectangle.pop(0)
            rectangle.pop(0)
            rectangle.pop(0)
    else:
        while len(rectangle) >= 3:
            a = int(rectangle[0])
            b = int(rectangle[1])
            c = int(rectangle[2])
            while a < c:
                rectangles.append([a, b, a + 1])
                a += 1
            rectangle.pop(0)
            rectangle.pop(0)
rectangles = sorted(rectangles, key=itemgetter(0, 2, 1))
result = []
previous_height = 0
for i in range(rectangles[0][0], rectangles[len(rectangles) - 1][0] + 1):
    max_height = 0
    for j in range(len(rectangles)):
        if rectangles[j][0] == i and rectangles[j][1] > max_height:
            max_height = rectangles[j][1]
    if max_height != previous_height:
        result.append(i)
        result.append(max_height)
    previous_height = max_height
result.append(rectangles[len(rectangles) - 1][2])
result = ','.join(map(str, result))
print(result)
